File: Bukkit/src/main/java/me/mrkirby153/kcutils/BossBar.java
package me.mrkirby153.kcutils;

import org.bukkit.Bukkit;
import org.bukkit.boss.BarColor;
import org.bukkit.boss.BarStyle;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.HashMap;
import java.util.UUID;

/**
 * A class for manipulating boss bars
 */
public class BossBar extends Module<JavaPlugin> implements Listener {

    private static HashMap<UUID, org.bukkit.boss.BossBar> playerBossBars = new HashMap<>();

    public BossBar(JavaPlugin plugin) {
        super("bossbar", "1.0", plugin);
    }

    public static org.bukkit.boss.BossBar getBar(Player player) {
        return playerBossBars.get(player.getUniqueId());
    }


    @EventHandler(ignoreCancelled = true)
    public void onPlayerQuit(PlayerQuitEvent event) {
        playerBossBars.remove(event.getPlayer().getUniqueId());
    }

    /**
     * Remove a boss bar from the player
     *
     * @param player The player
     */
    public void remove(Player player) {
        org.bukkit.boss.BossBar bar = playerBossBars.remove(player.getUniqueId());
        if (bar == null)
            return;
        bar.removeAll();
    }

    /**
     * Remove all online player's boss bars
     */
    public void removeAll() {
        Bukkit.getOnlinePlayers().forEach(this::remove);
    }

    /**
     * Sets a player's boss bar color
     *
     * @param player The player
     * @param color  The color of the bar
     */
    public void setColor(Player player, BarColor color) {
        org.bukkit.boss.BossBar bar = playerBossBars.get(player.getUniqueId());
        if (bar == null)
            return;
        bar.setColor(color);
    }

    /**
     * Set the boss bar's progress
     *
     * @param player  The player
     * @param percent The percent fill of the boss bar
     */
    public void setPercent(Player player, double percent) {
        org.bukkit.boss.BossBar bar = playerBossBars.get(player.getUniqueId());
        if (bar == null)
            return;
        if(percent < 0)
            percent = 0;
        if(percent > 1)
            percent = 1;
        bar.setProgress(percent);
    }

    /**
     * Set the boss bar's style
     *
     * @param player The player
     * @param style  The boss bar's style
     */
    public void setStyle(Player player, BarStyle style) {
        org.bukkit.boss.BossBar bar = playerBossBars.get(player.getUniqueId());
        if (bar == null)
            return;
        bar.setStyle(style);
    }

    /**
     * Set the boss bar's text
     *
     * @param player The player
     * @param text   The text
     */
    public void setTitle(Player player, String text) {
        org.bukkit.boss.BossBar bar = playerBossBars.get(player.getUniqueId());
        if ((text == null || text.isEmpty()) && bar != null) {
            bar.removeAll();
            playerBossBars.remove(player.getUniqueId());
            return;
        }
        if (bar == null) {
            playerBossBars.put(player.getUniqueId(), bar = Bukkit.createBossBar(text, BarColor.PINK, BarStyle.SOLID));
            bar.addPlayer(player);
        }
        bar.setTitle(text);
    }

    @Override
    protected void init() {
        getPlugin().getServer().getPluginManager().registerEvents(this, getPlugin());
    }
}

File: Bukkit/src/main/kotlin/me/mrkirby153/kcutils/BossBar.kt
package me.mrkirby153.kcutils

import org.bukkit.Bukkit
import org.bukkit.boss.BarColor
import org.bukkit.boss.BarStyle
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.plugin.java.JavaPlugin
import java.util.*

/**
 * A class for manipulating boss bars
 */
class BossBar(plugin: JavaPlugin) : Module<JavaPlugin>("bossbar", plugin), Listener {


    @EventHandler(ignoreCancelled = true)
    fun onPlayerQuit(event: PlayerQuitEvent) {
        playerBossBars.remove(event.player.uniqueId)
    }

    /**
     * Remove a boss bar from the player
     *
     * @param player The player
     */
    fun remove(player: Player) {
        val bar = playerBossBars.remove(player.uniqueId) ?: return
        bar.removeAll()
    }

    /**
     * Remove all online player's boss bars
     */
    fun removeAll() {
        Bukkit.getOnlinePlayers().forEach {
            remove(it)
        }
    }

    /**
     * Sets a player's boss bar color
     *
     * @param player The player
     * @param color  The color of the bar
     */
    fun setColor(player: Player, color: BarColor) {
        val bar = playerBossBars[player.uniqueId] ?: return
        bar.color = color
    }

    /**
     * Set the boss bar's progress
     *
     * @param player  The player
     * @param percent The percent fill of the boss bar
     */
    fun setPercent(player: Player, percent: Double) {
        val bar = playerBossBars[player.uniqueId] ?: return
        var prec = percent
        if (percent < 0)
            prec = 0.0
        if (percent > 1)
            prec = 1.0
        bar.progress = prec
    }

    /**
     * Set the boss bar's style
     *
     * @param player The player
     * @param style  The boss bar's style
     */
    fun setStyle(player: Player, style: BarStyle) {
        val bar = playerBossBars[player.uniqueId] ?: return
        bar.style = style
    }

    /**
     * Set the boss bar's text
     *
     * @param player The player
     * @param text   The text
     */
    fun setTitle(player: Player, text: String?) {
        var bar: org.bukkit.boss.BossBar? = playerBossBars[player.uniqueId]
        if ((text == null || text.isEmpty()) && bar != null) {
            bar.removeAll()
            playerBossBars.remove(player.uniqueId)
            return
        }
        if (bar == null) {
            bar = Bukkit.createBossBar(text, BarColor.PINK, BarStyle.SOLID)
            playerBossBars.put(player.uniqueId, bar)
            bar!!.addPlayer(player)
        }
        bar.title = text
    }

    override fun init() {
        plugin.server.pluginManager.registerEvents(this, plugin)
    }

    companion object {

        private val playerBossBars = HashMap<UUID, org.bukkit.boss.BossBar>()

        /**
         * Gets the boss bars for the player
         *
         * @param player The boss bar for the player
         */
        fun getBar(player: Player): org.bukkit.boss.BossBar? {
            return playerBossBars[player.uniqueId]
        }
    }
}
